source/ring_buffer.c

```diff
@@ -0,0 +1,205 @@
+        *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, requested_size);
+        return AWS_OP_SUCCESS;
+    }
+
+    /* you'll constantly bounce between the next two branches as the ring buffer is traversed. */
+    /* after N + 1 wraps */
+    if (tail_cpy > head_cpy) {
+        size_t space = tail_cpy - head_cpy - 1;
+
+        if (space >= requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + requested_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+        /* After N wraps */
+    } else if (tail_cpy < head_cpy) {
+        /* prefer the head space for efficiency. */
+        if ((size_t)(ring_buf->allocation_end - head_cpy) > requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + requested_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+
+        if ((size_t)(tail_cpy - ring_buf->allocation) > requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation + requested_size);
+            *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+    }
+
+    return aws_raise_error(AWS_ERROR_OOM);
+}
+
+int aws_ring_buffer_acquire_up_to(
+    struct aws_ring_buffer *ring_buf,
+    size_t minimum_size,
+    size_t requested_size,
+    struct aws_byte_buf *dest) {
+    AWS_ASSERT(requested_size >= minimum_size)
+
+    if (requested_size == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    uint8_t *tail_cpy = aws_atomic_load_ptr(&ring_buf->tail);
+    uint8_t *head_cpy = aws_atomic_load_ptr(&ring_buf->head);
+
+    /* this branch is, we don't have any vended buffers. */
+    if (head_cpy == tail_cpy) {
+        size_t ring_space = ring_buf->allocation_end - ring_buf->allocation;
+
+        size_t allocation_size = ring_space > requested_size ? requested_size : ring_space;
+
+        if (allocation_size < minimum_size) {
+            return aws_raise_error(AWS_ERROR_OOM);
+        }
+
+        /* go as big as we can. */
+        /* we don't have any vended, so this should be safe. */
+        aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation + allocation_size);
+        aws_atomic_store_ptr(&ring_buf->tail, ring_buf->allocation);
+        *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, allocation_size);
+        return AWS_OP_SUCCESS;
+    }
+    /* you'll constantly bounce between the next two branches as the ring buffer is traversed. */
+    /* after N + 1 wraps */
+    if (tail_cpy > head_cpy) {
+        size_t space = tail_cpy - head_cpy;
+        /* this shouldn't be possible. */
+        AWS_ASSERT(space);
+        space -= 1;
+
+        size_t returnable_size = space > requested_size ? requested_size : space;
+
+        if (returnable_size >= minimum_size) {
```

Comment:
If `minimum size == 0` and `tail_cpy - head_cpy == 1`, then `dest` will point to a buffer that points to NULL and has 0 capacity. Is this behaviour expected? PR #410 mentions this and adds a proof harness.

Reply:
will update in a future PR.